### ISSUE TITLE
Request: publish typeahead and bloodhound on npm

### ISSUE THREAD

**Issue by ragulka** [twitter/typeahead.js#743]
In the era of browserify and commonJS it makes sense to publish also front-end-only packages on npm. 
- npm stores the tagged versions on it's own servers
- when managing dependencies, i'd rather manage them all in npm (most of the dependencies I use are already on npm)
- it's nice to have the option, isn't it?

**Comment by clayzermk1** [comment]
> This should be improved once v0.11 lands as both bloodhound and the jquery plugin will be wrapped in https://github.com/umdjs/umd.

 +1

**Comment by unindented** [comment]
@jharding Sorry to insist, but with more than 150 open issues, and 37 pending pull requests, this repo is starting to look abandoned. Are there plans to revive it?

**Comment by jstejada** [comment]
+1

**Comment by santanubasu** [comment]
For anyone visiting this after the fact: the package on npmjs.com is literally called typeahead.js, not typeahead.  This probably should have been obvious to me, but when I searched for simply typeahead on npmjs I got a different module.  Not sure why both results don't come up...  Likewise, adding a dep on typeahead v0.11.1 in your package.json will fail, it needs to be "typeahead.js":"0.11.1".

**Comment by kylegoetz** [comment]
@santanubasu makes a very good point. I literally spent a whole day trying to figure out why the Typeahead I was using wasn't emitting the expected events, and then I discovered that Typeahead is a 1-year old version on NPM, while Typeahead**.js** is the correct, up-to-date version.

**Comment by kylegoetz** [comment]
Ah, as it turns out, it still doesn't work. The factory() in the NPM version doesn't return anything, so require('typeahead.js') returns undefined. I haven't figured out what it's supposed to return to work.

**Comment by santanubasu** [comment]
FWIW, I did require("typeahead.jquery.js") and that worked.  However, I didn't assign the result (undefined I assume) to any variable.  Instead, I just used it via jquery, for example: this.$el.typeahead(...)

**Comment by kylegoetz** [comment]
Thanks, @santanubasu. That's what I ended up doing, although personally I didn't like doing it that way. There should be a proper export so I can do Typeahead(inputElement, options) or something like that. Also I think I still had to make a change to the Typeahead JS file to get it working properly (maybe added a "return typeahead" or something at one point. I also had to require('jquery') in my own code or else it wouldn't work.

**Comment by ghost** [comment]
Including this library via npm is really annoying. I can include the library, but the module itself uses a `require('jquery')`... meaning it will pull its own instance of jquery, attach to `$.fn.typeahead`, and export nothing. So if I have somewhere else in my code `var $ = require ('jquery'); $(...).typeahead()` that particular `$` isn't going to have `fn.typeahead`. Anyone have a solution to this?

**Comment by santanubasu** [comment]
If you are only using browserify, you might be able to get around your issue using browserify-resolutions (https://github.com/Updater/browserify-resolutions).  I had the identical issue you reported with the bootstrap datetimepicker library, and this, um... resolved it.

Note that this doesn't address a situation where you have jquery as a global via some other means, and you also require() it from inside a browserify bundle.  That still seems to break in the way you reported, if anyone has a solution to that, would love to hear it.

**Comment by BrianGenisio** [comment]
@aft-luke YES!  I'm dealing with this exact problem.  Did you figure out a way around it?  I've spent all day trying to come up with a solution.

**Comment by ghost** [comment]
Yes - you can use a combination of browserify-shim and the package.json `browser: {}` with a `global` key to load jquery from a file (or a cdn), point browesrify to that global `jquery: 'global:jQuery'`, then all successive requires will actually be reaching into the singular jquery, which browserify pulled off of `window.jQuery`. More details [here](https://stackoverflow.com/questions/24835954/configure-a-generic-jquery-plugin-with-browserify-shim/25585778#25585778)

**Comment by nguyen-brian** [comment]
I'm in the same boat with the browserify people.  For a while my app was using the latest jquery version (2.1.3), so typeahead resolved its jquery dependency using the same version.  Then when 2.1.4 came out, doing a recompile would make typeahead have its own nested version of jquery separate from the one I was setting globally.  The solution was to update my app's version of jquery to 2.1.4 to make typeahead use the same version -- not optimal, but a huge relief once I understood what was happening.

**Comment by mikeymckay** [comment]
@nikvdp has packaged typeahead.js-browserify: https://www.npmjs.com/package/typeahead.js-browserify

It claims to solve these issues...trying it now!

**Comment by BrianGenisio** [comment]
@mikeymckay Let me know if you get it working.  Whenever I try to use this version, it brings in it's own version of jQuery.

I still haven't solved this problem in a way that I like.  I'm successfully telling Browserify to include the Bower Typeahead.js version with a Browserify shim, but I can't do it in a separated way.

By that, I mean, I can get Typeahead to use the global jQuery only if it is built into my app bundle.  But I like to have my app.js separate from my vendor.js.  For whatever reason, when I put typeahead.js into my vendor.js, Browserify simply won't use my global shim.  It is maddening.  So I have the Typeahead code being bundled with my app.js right now... not ideal, but at least it works.

**Comment by Ch4s3** [comment]
Has anyone else had any luck with this?

**Comment by mikeymckay** [comment]
My solution was to leave typeahead and use awesomeplete:
https://www.npmjs.com/package/awesomplete

**Comment by clayzermk1** [comment]
@Ch4s3 I am using substack/browserify and found it easy enough to do the following in my package.json both before and after @jharding published typeahead.js to npm.

``` javascript
{
  "browser": {
    "bloodhound": "./node_modules/typeahead.js/dist/bloodhound.js",
    "typeahead.js": "./node_modules/typeahead.js/dist/typeahead.jquery.js"
  },
  "dependencies": {
    "typeahead.js": "0.11.1"
    // or just go directly to github and bypass the npm registry
    "typeahead.js": "github:twitter/typeahead.js#0.11.1"
  }
}
```

**Comment by Helgesmeby** [comment]
This repo appears to be abandoned. Maintained fork: [https://github.com/corejavascript/typeahead.js](https://github.com/corejavascript/typeahead.js)

Also watch out for the npm typeahead library which is completely different from this.
